### ISSUE TITLE
Kpconv on scannet can not reach the performance in paper

### ISSUE THREAD

**Issue by jiazihankk** [nicolas-chaulet/torch-points3d#431]
I have tryed to run the command:
`poetry run python train.py training=scannet_benchmark/kpconv task=segmentation model_type=kpconv model_name=KPConvPaper dataset=scannet`
to test kpconv on scannet dataset, but can not reach the performance in paper. the mIou on valid set in paper is around 68
https://arxiv.org/abs/1904.08889
but I only get 57 finally, detail:
https://app.wandb.ai/jamesjian003/scannet-benchmark/runs/36suzt8u?workspace=user-jamesjian003
i wonder what's problem in it ?

**Comment by nicolas-chaulet** [comment]
Hi! It is likely to come from the way we handle the dataset. KPConv original codebase has a clever strategy to sample the data, it takes 2m radius spheres from each sample with more weights being given to the more rare classes. This and a voting inference scheme boosts performance quite a bit and could explain the discrepancy. We have followed their strategy on S3DIS and could reproduce their results. We also use a 5cm grid (as opposed to 4cm in the paper). I think we achieved 60% with those settings, let me find the logs.

**Comment by loicland** [comment]
[This run](https://wandb.ai/loicland/scannet-benchmark/runs/2l26ksb3?workspace=user-loicland) reached $60.5$% mIoU on the valdiation set.

**Comment by rohitgirdhar** [comment]
Hi @nicolas-chaulet, can you share the setup used to reproduce the results for S3DIS? I am using the test script to evaluate the pre-trained models (`python scripts/cv_s3dis.py checkpoint_dir=`pwd` model_name=kpconv`). What value do you set `voting_runs` to?

**Comment by rohitgirdhar** [comment]
Hi @nicolas-chaulet, @tchaton 
Would really appreciate any pointers on how to repro the S3DIS performance. Increasing the `voting_runs` didn't seem to have any effect as it seems to sample the exact same spheres in each voting run. It seems the default grid sampling only samples around 7% of the point cloud, which is probably what is causing the low performance, but I've not been able to figure out how to increase the number of points being sampled (it seems it is supposed to sample every possible grid [here](https://github.com/nicolas-chaulet/torch-points3d/blob/9966f2350e03165158ba40b9f203aae7a16d31aa/torch_points3d/datasets/segmentation/s3dis.py#L572)?) Here's how I'm running it (changed `cv_s3dis.py` to only run the area 5 evaluation), and the output I get is as follows (expecting ~67% mIOU for area 5 evaluation, as per the paper):
```
$ python scripts/cv_s3dis.py checkpoint_dir=`pwd` model_name=kpconv cuda=0
...
[2021-02-22 11:31:44,423][torch_points3d.trainer][INFO] - Dataset: S3DISFusedDataset
train_pre_batch_collate_transform = None
val_pre_batch_collate_transform = None
test_pre_batch_collate_transform = None
pre_transform = None
test_transform = Compose([
    XYZFeature(axis=['z']),
    AddFeatsByKeys(rgb=True, pos_z=True),
    Center(),
    MultiScaleTransform,
])
train_transform = Compose([
    RandomNoise(sigma=0.001, clip=0.05),
    RandomRotate((-180, 180), axis=2),
    RandomScaleAnisotropic([0.8, 1.2]),
    Random symmetry of axes: x=True, y=False, z=False,
    DropFeature: proba = 0.2, feature = rgb,
    XYZFeature(axis=['z']),
    AddFeatsByKeys(rgb=True, pos_z=True),
    Center(),
    MultiScaleTransform,
])
val_transform = Compose([
    XYZFeature(axis=['z']),
    AddFeatsByKeys(rgb=True, pos_z=True),
    Center(),
    MultiScaleTransform,
])
inference_transform = Compose([
    XYZFeature(axis=['z']),
    AddFeatsByKeys(rgb=True, pos_z=True),
    Center(),
])
pre_collate_transform = Compose([
    PointCloudFusion(),
    SaveOriginalPosId,
    GridSampling3D(grid_size=0.04, quantize_coords=False, mode=mean),
])
Size of train_dataset = 3000
Size of test_dataset = 1679
Size of val_dataset = 1166
Batch size = 1
[2021-02-22 11:31:44,539][torch_points3d.datasets.base_dataset][INFO] - Available stage selection datasets:  ['test', 'val']
[2021-02-22 11:31:44,539][torch_points3d.datasets.base_dataset][INFO] - The models will be selected using the metrics on following dataset:  None
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [04:03<00:00,  6.90it/s, test_acc=84.68, test_loss_seg=0.698, test_macc=67.33, test_miou=59.75)]
[2021-02-22 11:35:57,795][torch_points3d.metrics.s3dis_tracker][INFO] - Computing full res mIoU, we have predictions for 6.74% of the points.
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] - ==================================================
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_loss_seg = 0.6984611661517299
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_acc = 84.68783830620407
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_macc = 67.33965057983178
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_miou = 59.759928082954
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_miou_per_class = {0: '90.02', 1: '97.98', 2: '75.69', 3: '0.00', 4: '13.01', 5: '49.44', 6: '45.39', 7: '87.05', 8: '72.31', 9: '71.08', 10: '65.33', 11: '55.11', 12: '54.46'}
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_iou_per_class = {'ceiling': 0.900154559613416, 'floor': 0.9797712478828235, 'wall': 0.7569019421103846, 'beam': 1e-08, 'column': 0.13012442466130214, 'window': 0.4943957772977914, 'door': 0.45394490175925345, 'chair': 0.870505324575796, 'table': 0.7231438413648141, 'bookcase': 0.7108385953439388, 'sofa': 0.6532736501461152, 'board': 0.5511392829134512, 'clutter': 0.5445970931149334}
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_full_vote_miou = 60.541203109757525
[2021-02-22 11:37:27,054][torch_points3d.metrics.base_tracker][INFO] -     test_vote_miou = 60.39961056864733
[2021-02-22 11:37:27,055][torch_points3d.metrics.base_tracker][INFO] - ==================================================
[2021-02-22 11:37:27,480][__main__][INFO] - ====================================================
[2021-02-22 11:37:27,480][__main__][INFO] - NUM POINTS : 78649818
[2021-02-22 11:37:27,480][__main__][INFO] - OA: 86.14103976693247
[2021-02-22 11:37:27,480][__main__][INFO] - MACC: 67.85410925295365
[2021-02-22 11:37:27,480][__main__][INFO] - MIOU : 60.541203109757525
[2021-02-22 11:37:27,480][__main__][INFO] - ====================================================
```

**Comment by nicolas-chaulet** [comment]
This looks a bit low indeed. You should get 62-63% on the best test epoch with the following training script: 
https://github.com/nicolas-chaulet/torch-points3d/blob/master/conf/training/s3dis_benchmark/kpconv.yaml

Also, we haven't implemented the deformable version so the reported performance is 65.4% and to achieve that you need to have voting activated as well as the data augmentation transforms even at test time. I hope this helps,
Best,
Nicolas

**Comment by rohitgirdhar** [comment]
Thanks @nicolas-chaulet, I assume you mean copying over all train-time augmentations (with the same parameters) to test time, which should presumably lead to different predictions across voting runs and then voting across those? 
- Could you share how many `voting_runs=` did you use? 
- Which s3dis dataset version do you use? `s3disfused`, `s3disfused-fixed` or `s3disfused-sparse`? I assume I'll have to edit the `test_transform:` part and copy over the train transforms and then train the model?
- Also, could you confirm if it's expected to only get predictions for 6.74% of the points?

Thanks again

**Comment by nicolas-chaulet** [comment]
Hi,
- 10 voting runs should do just fine
- `test_full_vote_miou` is actually the miou on all points (after upsampling)
- `s3disfused.yaml` should be the one. And yes it is about copying the train transforms other to the test transforms, you will have to do that in code there: https://github.com/nicolas-chaulet/torch-points3d/blob/9966f2350e03165158ba40b9f203aae7a16d31aa/torch_points3d/datasets/segmentation/s3dis.py#L663

**Comment by rohitgirdhar** [comment]
Thanks @nicolas-chaulet, let me try that. Yes I understand `test_full_vote_miou` is after upsampling, my concern is more about this statement `[2021-02-22 12:40:28,999][torch_points3d.metrics.s3dis_tracker][INFO] - Computing full res mIoU, we have predictions for 6.74% of the points`, which suggest we are doing forward pass on only 6.74% of the points and getting the rest via upsampling. I wonder if there's a way to sample the spheres more densely, and if you found that does not help?

**Comment by rohitgirdhar** [comment]
I replaced the transforms for val/test to be same as train in [here](https://github.com/nicolas-chaulet/torch-points3d/blob/9966f2350e03165158ba40b9f203aae7a16d31aa/torch_points3d/datasets/segmentation/s3dis.py#L663) like so:
```
        self.val_dataset = dataset_cls(
            self._data_path,
            sample_per_epoch=-1,
            test_area=self.dataset_opt.fold,
            split="val",
            pre_collate_transform=self.pre_collate_transform,
            transform=self.train_transform,
        )
        self.test_dataset = dataset_cls(
            self._data_path,
            sample_per_epoch=-1,
            test_area=self.dataset_opt.fold,
            split="test",
            pre_collate_transform=self.pre_collate_transform,
            transform=self.train_transform,
        )
```
Testing with 10 voting runs for the provided fold-5 model unfortunately didn't lead to any improvements, got 59.5 full vote mIOU

```
Size of train_dataset = 3000
Size of test_dataset = 1679
Size of val_dataset = 1166
Batch size = 1
[2021-02-22 15:11:34,750][torch_points3d.datasets.base_dataset][INFO] - Available stage selection datasets:  ['test', 'val']
[2021-02-22 15:11:34,751][torch_points3d.datasets.base_dataset][INFO] - The models will be selected using the metrics on following dataset:  None
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:35<00:00,  4.24it/s, test_acc=83.44, test_loss_seg=0.751, test_macc=65.13, test_miou=56.93)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:33<00:00,  4.27it/s, test_acc=83.45, test_loss_seg=0.749, test_macc=65.17, test_miou=57.05)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:31<00:00,  4.29it/s, test_acc=83.45, test_loss_seg=0.748, test_macc=65.20, test_miou=57.08)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:31<00:00,  4.29it/s, test_acc=83.45, test_loss_seg=0.749, test_macc=65.26, test_miou=57.13)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:30<00:00,  4.30it/s, test_acc=83.46, test_loss_seg=0.748, test_macc=65.31, test_miou=57.17)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:29<00:00,  4.31it/s, test_acc=83.46, test_loss_seg=0.748, test_macc=65.30, test_miou=57.15)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:31<00:00,  4.28it/s, test_acc=83.44, test_loss_seg=0.749, test_macc=65.29, test_miou=57.11)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:30<00:00,  4.30it/s, test_acc=83.44, test_loss_seg=0.749, test_macc=65.32, test_miou=57.17)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:30<00:00,  4.30it/s, test_acc=83.45, test_loss_seg=0.749, test_macc=65.34, test_miou=57.19)]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1679/1679 [06:30<00:00,  4.30it/s, test_acc=83.45, test_loss_seg=0.749, test_macc=65.35, test_miou=57.21)]
[2021-02-22 16:16:59,152][torch_points3d.metrics.s3dis_tracker][INFO] - Computing full res mIoU, we have predictions for 6.74% of the points.
[2021-02-22 16:18:24,193][torch_points3d.metrics.base_tracker][INFO] - ==================================================
[2021-02-22 16:18:24,193][torch_points3d.metrics.base_tracker][INFO] -     test_loss_seg = 0.7492597097026185
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_acc = 83.45636776642111
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_macc = 65.35345498558846
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_miou = 57.21747760211273
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_miou_per_class = {0: '89.63', 1: '97.88', 2: '73.93', 3: '0.02', 4: '11.94', 5: '43.98', 6: '43.63', 7: '85.68', 8: '70.93', 9: '68.79', 10: '62.35', 11: '42.41', 12: '52.66'}
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_iou_per_class = {'ceiling': 0.8962922593915846, 'floor': 0.9787991559786914, 'wall': 0.7392999019686104, 'beam': 0.0001952510553788867, 'column': 0.1194483956588123, 'window': 0.43981569263154063, 'door': 0.43633178248889104, 'chair': 0.8568286909268553, 'table': 0.7092524777788098, 'bookcase': 0.687870900637449, 'sofa': 0.6234697286549784, 'board': 0.42408708814718155, 'clutter': 0.5265807629558729}
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_full_vote_miou = 59.538073947497175
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] -     test_vote_miou = 59.48456356180773
[2021-02-22 16:18:24,194][torch_points3d.metrics.base_tracker][INFO] - ==================================================
[2021-02-22 16:18:24,498][__main__][INFO] - ====================================================
[2021-02-22 16:18:24,499][__main__][INFO] - NUM POINTS : 78649818
[2021-02-22 16:18:24,499][__main__][INFO] - OA: 85.98778194248332
[2021-02-22 16:18:24,499][__main__][INFO] - MACC: 66.7603655442342
[2021-02-22 16:18:24,499][__main__][INFO] - MIOU : 59.538073947497175
[2021-02-22 16:18:24,499][__main__][INFO] - ====================================================
```

**Comment by nicolas-chaulet** [comment]
Hey, Yes, I forgot to add that you would also have to set the number of
sample per epochs in the test set to something like 2000. This will ensure
that spheres are sampled randomly in the space (as opposed to on a fixed
grid). Also please set `weight_name=miou` to select best model on train for
iou. If you have your wandb logs feel free to share them with me, I can
take a look and see if everything is normal. Also, you might want to train
longer (500 epochs) this will stabilise the model. For the benchmark in our
paper we wanted fair evaluation and 500 epochs is a lot for the other
models so we took that down a bit. Getting back the perf on KPConv was a
little tricky but we got there eventually (combination of training long
enough, correct voting and data augmentation at test time did the trick).

Here are some older logs that might help you, you'll see the performance as
well as all the config for the model and dataset.
https://wandb.ai/nicolas/s3dis/runs/2x0vuham/logs?workspace=user-nicolas

Le mar. 23 févr. 2021 à 00:25, Rohit Girdhar <notifications@github.com> a
écrit :

**Comment by bxiang233** [comment]
Hi all, has any of you achieved the reported performance (~65.4%) in S3DIS dataset based on KPConv backbone? Could you please share me with the exact configuration files you have? Thanks a lot!

**Comment by CCInc** [comment]
@bxiang233 Did you take a look at the run Nicolas posted above? You can see it reaches a result of 65.66 miou on the validation step. The config files are included in the run.https://wandb.ai/nicolas/s3dis/runs/2x0vuham/overview

**Comment by bxiang233** [comment]
> @bxiang233 Did you take a look at the run Nicolas posted above? You can see it reaches a result of 65.66 miou on the validation step. The config files are included in the run.https://wandb.ai/nicolas/s3dis/runs/2x0vuham/overview

Thanks a lot!
